File: backend/app/utils/telegram.py
import httpx
from typing import Optional, Dict, Any
from app.core.config import settings


async def validate_telegram_token(token: str) -> Optional[Dict[str, Any]]:
    """Валидирует Telegram bot token и возвращает информацию о боте"""
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                f"{settings.TELEGRAM_API_URL}/bot{token}/getMe"
            )
            if response.status_code == 200:
                data = response.json()
                if data.get("ok"):
                    return data.get("result")
            return None
    except Exception:
        return None


async def get_bot_info(token: str) -> Optional[Dict[str, Any]]:
    """Получает информацию о боте по токену"""
    return await validate_telegram_token(token)


def normalize_channel_url(channel_input: str) -> str:
    """Нормализует ввод канала в полный URL"""
    if not channel_input:
        return channel_input
    
    channel_input = channel_input.strip()
    
    # Если уже полный URL, возвращаем как есть
    if channel_input.startswith(("https://t.me/", "http://t.me/", "https://", "http://")):
        return channel_input
    
    # Если формат @username, преобразуем в https://t.me/username
    if channel_input.startswith("@"):
        username = channel_input[1:]  # Убираем @
        return f"https://t.me/{username}"
    
    # Если просто username без @, добавляем префикс
    if channel_input and not channel_input.startswith(("http", "@")):
        return f"https://t.me/{channel_input}"
    
    return channel_input


async def get_channel_info(channel_url: str) -> Optional[str]:
    """Пытается получить название канала из URL или username"""
    try:
        if not channel_url:
            return None
        
        # Нормализуем URL
        normalized_url = normalize_channel_url(channel_url)
        
        # Если формат @username, извлекаем username
        if channel_url.startswith("@"):
            username = channel_url[1:].strip()
            if username:
                # Преобразуем в читаемое название
                readable_name = username.replace("_", " ").replace("-", " ")
                readable_name = " ".join(word.capitalize() for word in readable_name.split())
                return readable_name
        
        # Парсим URL канала Telegram
        # Формат: https://t.me/channel_name или https://t.me/joinchat/...
        if not normalized_url.startswith(("https://t.me/", "http://t.me/")):
            return None
        
        # Извлекаем имя канала из URL
        parts = normalized_url.replace("https://t.me/", "").replace("http://t.me/", "").split("/")
        if parts:
            channel_name = parts[0].strip()
            # Убираем параметры запроса
            if "?" in channel_name:
                channel_name = channel_name.split("?")[0]
            if channel_name and channel_name != "joinchat":
                # Преобразуем в читаемое название (убираем подчеркивания, делаем заглавной первую букву)
                readable_name = channel_name.replace("_", " ").replace("-", " ")
                readable_name = " ".join(word.capitalize() for word in readable_name.split())
                return readable_name
    except Exception:
        pass
    return None

File: backend-node/src/utils/telegram.ts
import { config } from '../core/config';

export interface TelegramBotInfo {
  id: number;
  is_bot: boolean;
  first_name: string;
  username?: string;
  can_join_groups?: boolean;
  can_read_all_group_messages?: boolean;
  supports_inline_queries?: boolean;
}

export interface TelegramApiResponse<T> {
  ok: boolean;
  result?: T;
  description?: string;
}

/**
 * Валидирует Telegram bot token и возвращает информацию о боте
 */
export async function validateTelegramToken(
  token: string
): Promise<TelegramBotInfo | null> {
  try {
    const response = await fetch(
      `${config.telegramApiUrl}/bot${token}/getMe`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
        signal: AbortSignal.timeout(10000), // 10 секунд таймаут
      }
    );

    if (!response.ok) {
      return null;
    }

    const data = await response.json() as TelegramApiResponse<TelegramBotInfo>;

    if (data.ok && data.result) {
      return data.result;
    }

    return null;
  } catch (error) {
    console.error('Error validating Telegram token:', error);
    return null;
  }
}

/**
 * Получает информацию о боте по токену (алиас для validateTelegramToken)
 */
export async function getBotInfo(
  token: string
): Promise<TelegramBotInfo | null> {
  return validateTelegramToken(token);
}

/**
 * Нормализует ввод канала в полный URL
 */
export function normalizeChannelUrl(channelInput: string): string {
  if (!channelInput) {
    return channelInput;
  }

  channelInput = channelInput.trim();

  // Если уже полный URL, возвращаем как есть
  if (
    channelInput.startsWith('https://t.me/') ||
    channelInput.startsWith('http://t.me/') ||
    channelInput.startsWith('https://') ||
    channelInput.startsWith('http://')
  ) {
    return channelInput;
  }

  // Если формат @username, преобразуем в https://t.me/username
  if (channelInput.startsWith('@')) {
    const username = channelInput.slice(1); // Убираем @
    return `https://t.me/${username}`;
  }

  // Если просто username без @, добавляем префикс
  if (channelInput && !channelInput.startsWith('http') && !channelInput.startsWith('@')) {
    return `https://t.me/${channelInput}`;
  }

  return channelInput;
}

/**
 * Пытается получить название канала из URL или username
 */
export async function getChannelInfo(
  channelUrl: string
): Promise<string | null> {
  try {
    if (!channelUrl) {
      return null;
    }

    // Нормализуем URL
    const normalizedUrl = normalizeChannelUrl(channelUrl);

    // Если формат @username, извлекаем username
    if (channelUrl.startsWith('@')) {
      const username = channelUrl.slice(1).trim();
      if (username) {
        // Преобразуем в читаемое название
        const readableName = username
          .replace(/_/g, ' ')
          .replace(/-/g, ' ');
        return readableName
          .split(' ')
          .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
          .join(' ');
      }
    }

    // Парсим URL канала Telegram
    // Формат: https://t.me/channel_name или https://t.me/joinchat/...
    if (
      !normalizedUrl.startsWith('https://t.me/') &&
      !normalizedUrl.startsWith('http://t.me/')
    ) {
      return null;
    }

    // Извлекаем имя канала из URL
    const parts = normalizedUrl
      .replace('https://t.me/', '')
      .replace('http://t.me/', '')
      .split('/');
    if (parts.length > 0) {
      let channelName = parts[0].trim();
      // Убираем параметры запроса
      if (channelName.includes('?')) {
        channelName = channelName.split('?')[0];
      }
      if (channelName && channelName !== 'joinchat') {
        // Преобразуем в читаемое название
        const readableName = channelName.replace(/_/g, ' ').replace(/-/g, ' ');
        return readableName
          .split(' ')
          .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
          .join(' ');
      }
    }
  } catch (error) {
    console.error('Error getting channel info:', error);
  }
  return null;
}
